### PR TITLE
Use asynchronous Azure storage APIs

### PR DESCRIPTION
**Pull request by Arun-LinkedIn**
This change will use asynchronous Azure storage APIs in code but will block on the response until rest of the cloud store code is refactored to handle responses asynchronously.

### CODE REVIEW

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ADAuthBasedStorageClient.java [hunk 1]
```diff
@@ -170,17 +171,50 @@ protected boolean handleExceptionAndHintRetry(BlobStorageException blobStorageEx
...
+      scheduledFutureRef.set(tokenRefreshScheduler.schedule(this::refreshTokenAndStorageClients,
+          (long) ((accessTokenRef.get().getExpiresAt().toEpochSecond() - OffsetDateTime.now().toEpochSecond())
+              * azureCloudConfig.azureStorageClientRefreshFactor), TimeUnit.SECONDS));
+    } catch (MalformedURLException | InterruptedException | ExecutionException ex) {
+      logger.error("Error building ABS blob service client: {}", ex.getMessage());
+      throw new IllegalStateException(ex);
+    }
+  }
+
+  private void refreshToken() throws MalformedURLException, ExecutionException, InterruptedException {
```
Comment (ankagrawal) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ADAuthBasedStorageClient.java:210]: javadocs?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/StorageClient.java [hunk 2]
```diff
@@ -353,6 +617,33 @@ protected BlobServiceClient createBlobStorageClient() {
     }
   }
 
+  /**
+   * Create the {@link BlobServiceAsyncClient} object.
+   * @param {@link CloudConfig} object.
```
Comment (ankagrawal) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/StorageClient.java:622]: maybe this javadoc needs to be updated.

**Review by ankagrawal — APPROVED** [review]
Looks good to me overall. Please fix some of the minor javadocs comments.

**Review by cgtz — COMMENTED** [review]
Looks good in its current form. One question about if this PR is meant to define the API for eventual async users or mostly to test usage of the async azure client in existing flows.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/StorageClient.java [hunk 3]
```diff
@@ -164,11 +184,73 @@ public BlobServiceClient getStorageClient() {
...
+   * @param tier {@link AccessTier} for the destination blob.
+   * @param contentMd5 An MD5 hash of the block content.
+   * @param requestConditions {@link BlobRequestConditions}
+   * @param timeout An optional timeout value beyond which a {@link RuntimeException} will be raised.
+   * @return The information of the uploaded block blob.
+   * @throws UnexpectedLengthException when the length of data does not match the input {@code length}.
+   * @throws NullPointerException if the input data is null.
+   * @throws UncheckedIOException If an I/O error occurs
+   */
+  public Response<BlockBlobItem> uploadAsyncWithResponse(BlobId blobId, InputStream data, long length,
```
[the remark this answers is not in the document]
Comment (cgtz) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/azure/StorageClient.java:216]: Made the decision to keep the scope limited in this PR.

**Pull request merged**